pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,462 @@
+	endWithLookaround := end + i.Config.QueryLookaroundPeriod.Milliseconds()
+
+	blocks := make([]*metastorev1.BlockMeta, 0)
+
+	for _, meta := range i.allPartitions { // TODO aleks-p: consider using binary search to find a good starting point
+		if meta.overlaps(startWithLookaround, endWithLookaround) {
+			p := i.getPartition(meta)
+			tenantBlocks := i.collectTenantBlocks(p, start, end, tenants)
+			blocks = append(blocks, tenantBlocks...)
+		}
+	}
+
+	return blocks, nil
+}
+
+func (i *Index) sortPartitions() {
+	slices.SortFunc(i.allPartitions, func(a, b *PartitionMeta) int {
+		return a.compare(b)
+	})
+}
+
+func (i *Index) collectTenantBlocks(p *indexPartition, start, end int64, tenants map[string]struct{}) []*metastorev1.BlockMeta {
+	blocks := make([]*metastorev1.BlockMeta, 0)
+	for _, s := range p.shards {
+		for tKey, t := range s.tenants {
+			_, ok := tenants[tKey]
+			if !ok && tKey != "" {
+				continue
+			}
+			for _, block := range t.blocks {
+				if start < block.MaxTime && end >= block.MinTime {
+					blocks = append(blocks, block)
+				}
+			}
+		}
+	}
+	return blocks
+}
+
+// ReplaceBlocks removes source blocks from the index and inserts replacement blocks into the index. The intended usage
+// is for block compaction. The replacement blocks could be added to the same or a different partition.
+func (i *Index) ReplaceBlocks(sources []string, sourceShard uint32, sourceTenant string, replacements []*metastorev1.BlockMeta) map[string]*BlockWithPartition {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	for _, newBlock := range replacements {
+		i.insertBlock(newBlock)
+	}
+
+	deletedBlocks := make(map[string]*BlockWithPartition, len(sources))
+	for _, sourceBlock := range sources {
+		b, meta := i.deleteBlock(sourceShard, sourceTenant, sourceBlock)
+		if b != nil && meta != nil {
+			deletedBlocks[sourceBlock] = &BlockWithPartition{
+				Meta:  meta,
+				Block: b,
+			}
+		}
+	}
+
+	return deletedBlocks
+}
+
+// deleteBlock deletes a block from the index. It is the caller's responsibility to enforce safe concurrent access.
+func (i *Index) deleteBlock(shard uint32, tenant string, blockId string) (*metastorev1.BlockMeta, *PartitionMeta) {
+	// first try the currently mapped partition
+	key := i.CreatePartitionKey(blockId)
+	if b, meta, ok := i.tryDelete(key, shard, tenant, blockId); ok {
+		return b, meta
+	}
+
+	// now try all other possible partitions
+	t := ulid.Time(ulid.MustParse(blockId).Time()).UTC().UnixMilli()
+
+	for _, p := range i.allPartitions {
+		if p.contains(t) {
+			if b, meta, ok := i.tryDelete(p.Key, shard, tenant, blockId); ok {
+				return b, meta
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func (i *Index) tryDelete(key PartitionKey, shard uint32, tenant string, blockId string) (*metastorev1.BlockMeta, *PartitionMeta, bool) {
+	meta := i.findPartitionMeta(key)
+	if meta == nil {
+		return nil, nil, false
+	}
+
+	p := i.getPartition(meta)
+
```

Comment:
I think we shouldn't load partition from disk here:
1. If partition (its part you want to modify) is memory – update it.
2. Delete the block from store.

Reply:
Agreed, we still need to find partition for the block in order to delete it but there is no need to load the partition.

Reply:
This is now removed and the delete contract is simplified. We no longer return back information about the partition the block was removed from. As a consequence, we now try to remove the block from several partitions in the store, however this only happens when the partition duration changes.